src/material-experimental/form-field/testing/form-field-harness.ts

```diff
@@ -0,0 +1,231 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  ComponentHarness,
+  HarnessEnvironment,
+  HarnessPredicate,
+  TestElement
+} from '@angular/cdk/testing';
+import {MatInputHarness} from '@angular/material-experimental/input/testing';
+// TODO(devversion): we cannot import the select harness through the module name
+// because it does not expose an entry-point that re-exports the harness
+import {MatSelectHarness} from '../../mdc-select/harness/select-harness';
+import {FormFieldHarnessFilters} from './form-field-harness-filters';
+
+// TODO(devversion): support datepicker harness once developed (COMP-203).
+// Also support chip list harness.
+/** Possible harnesses of controls which can be bound to a form-field. */
+export type FormFieldControlHarness = MatInputHarness|MatSelectHarness;
+
+/**
+ * Harness for interacting with a standard Material form-field's in tests.
+ * @dynamic
+ */
+export class MatFormFieldHarness extends ComponentHarness {
+  static hostSelector = '.mat-form-field';
+
+  /**
+   * Gets a `HarnessPredicate` that can be used to search for an form-field with
+   * specific attributes.
+   * @param options Options for narrowing the search:
+   *   - `selector` finds a form-field that matches the given selector.
+   * @return a `HarnessPredicate` configured with the given options.
+   */
+  static with(options: FormFieldHarnessFilters = {}): HarnessPredicate<MatFormFieldHarness> {
+    return new HarnessPredicate(MatFormFieldHarness, options);
+  }
+
+  private _prefixContainer = this.locatorForOptional('.mat-form-field-prefix');
+  private _suffixContainer = this.locatorForOptional('.mat-form-field-suffix');
+  private _label = this.locatorForOptional('.mat-form-field-label');
+  private _errors = this.locatorForAll('.mat-error');
+  private _hints = this.locatorForAll('mat-hint,.mat-hint');
+
+  private _controlHarnessSelector =
+      [MatInputHarness, MatSelectHarness].map(h => h.hostSelector).join(',');
+
+  /** Gets the appearance of the form-field. */
+  async getAppearance(): Promise<'legacy'|'standard'|'fill'|'outline'> {
+    const hostEl = await this.host();
+    const [isLegacy, isStandard, isFill, isOutline] = await Promise.all([
+      hostEl.hasClass('mat-form-field-appearance-legacy'),
+      hostEl.hasClass('mat-form-field-appearance-standard'),
+      hostEl.hasClass('mat-form-field-appearance-fill'),
+      hostEl.hasClass('mat-form-field-appearance-outline'),
+    ]);
+    if (isLegacy) {
+      return 'legacy';
+    } else if (isStandard) {
+      return 'standard';
+    } else if (isFill) {
+      return 'fill';
+    } else if (isOutline) {
+      return 'outline';
+    }
```

Comment:
What do you think of condensing this to
```ts
const hostClass = await (await this.host()).getAttribute('class');
const appearanceMatch = hostClass.match(/mat-form-field-appearance-(\w+)/);
if (appearanceMatch) {
  return appearanceMatch[1];
}

throw Error('Could not determine appearance of form-field.');
```

Reply:
I like that it is way shorter and clean. Though it's not 100% accurate as it could also match accidental cases where the class looks like `mat-form-field-appearance-custom`.

We could fix that by adding the valid options to the capture group I guess.

Reply:
I think this regex should get just the cases we want:
`/mat-form-field-appearance-(legacy|standard|fill|outline)(?:$|\s)/`

Reply:
:+1: and good point with matching either end or a space!

Reply:
Done.